### ISSUE TITLE
[BUG] Error while trying to publish events using Event Grid's publishCloudEvent (sync) and using jar 4.17.0-beta.1

### ISSUE THREAD

**Issue by 1-javier-fernandez** [Azure/azure-sdk-for-java#37039]
**Describe the bug**
The following root cause exception is thrown when attempting to publish an event using Event Grid's synchronous EventGridClient.publishCloudEvent in a SpringBoot 2.7.3 application:

java.lang.IllegalStateException: block()/blockFirst()/blockLast() are blocking, which is not supported in thread reactor-http-epoll-2

Note: the same code works as expected using a project with just the azure-messaging-eventgrid and azure-core dependencies.

***Exception or Stack Trace***

Stack trace is too long. I have attached it.

**To Reproduce**

Create a Spring Boot application version 2.7.3 that uses Event Grid's 4.17.0-beta.1 jar. Then use code like in the coming section.

***Code Snippet***

   public static final String TOPICNAME = "jfgTpc1";
    public static final String EVENT_SUBSCRIPTION_NAME = "evtsub1";
    public static final String ENDPOINT = "";  // namespace endpoint
    public static final AzureKeyCredential CREDENTIAL = new AzureKeyCredential("<key for namespace  http endpoint>");

  public void publishCloudEventSync() {
        EventGridClient client = buildClient();
        PublishResult publishResult = client.publishCloudEvent(TOPICNAME, getCloudEvent());
    }

    EventGridClient buildClient() {
        return buildClientBuilder().buildClient();
    }

    EventGridAsyncClient buildAsyncClient() {
        return buildClientBuilder().buildAsyncClient();
    }

    EventGridClientBuilder buildClientBuilder() {
        return new EventGridClientBuilder()
                .httpClient(HttpClient.createDefault())
                .endpoint(ENDPOINT)
                .serviceVersion(EventGridMessagingServiceVersion.V2023_06_01_PREVIEW)
                .credential(CREDENTIAL);
    }

    private static CloudEvent getCloudEvent() {
        return new CloudEvent("/events/testEvent", "Microsoft.MockPublisher.TestEvent",
                BinaryData.fromObject(new HashMap<String, String>() {
                    {
                        put("Field1", "Value1");
                        put("Field2", "Value2");
                        put("Field3", "Value3");
                    }
                }), CloudEventDataFormat.JSON, "application/json")
                .setSubject("Test")
                .setTime(OffsetDateTime.now());
    }

**Expected behavior**
The code should send 1 event to the namespace topic configured. You should be able to see on the namespace's metrics (Azure portal and on the namespace overview page) that event(s) have been published successfully.

**Screenshots**
N/A

**Setup (please complete the following information):**
 - OS: Ubuntu jammy (Ubuntu 22.04.3 LTS)
 - IDE:IntelliJ IDEA 2023.2.2     Build #IC-232.9921.47, built on September 12, 2023
 - Library/Libraries: 
 com.azure:azure-sdk-bom:1.2.16
com.azure:azure-messaging-eventgrid:4.17.0-beta.1
org.projectlombok:lombok:1.18.28
org.springframework.boot:spring-boot-starter-actuator
org.springframework.boot:spring-boot-starter-webflux
org.assertj:assertj-core
org.junit.jupiter:junit-jupiter-api
  
 - Java version: 1.8
 - App Server/Environment: Embeded web server in Spring Boot. That embded server should be Reactor Netty because I am using org.springframework.boot:spring-boot-starter-webflux
 - 
 - Frameworks: Spring Boot 2.7.3

**Additional context**

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [X ] Bug Description Added
- [ ] Repro Steps Added
- [ ] Setup information Added

**Comment by 1-javier-fernandez** [comment]
You may tag me by also using @jfggdl (I am the Event Grid Product Manager).

**Comment by 1-javier-fernandez** [comment]
[stack trace.txt](https://github.com/Azure/azure-sdk-for-java/files/12799223/stack.trace.txt)

**Comment by 1-javier-fernandez** [comment]
[simple-project-that-works-left-vs-my-project-dependencies-right.pdf](https://github.com/Azure/azure-sdk-for-java/files/12805126/simple-project-that-works-left-vs-my-project-dependencies-right.pdf)

**Comment by 1-javier-fernandez** [comment]
I have added a comparison report that shows the dependencies of a simple project that depends only on the messaging.eventgrid jar, which it works as expected, vs my project's dependencies. There are minor version changes between the two set of dependencies where they have common jars.

**Comment by billwert** [comment]
@saragluna FYI - I will take a first look but if you see something that jumps out please speak up! :)

**Comment by anuchandy** [comment]
@billwert - Just sharing some thoughts. One possibility is, the call site of `publishCloudEventSync()` happens to run on Netty-IO thread. For example, if call site is a callback that Netty-Server calls upon the arrival HTTP request (mapped to some REST Path).

If the SDK sync API (sync over async) underneath blocks the current thread and given the current thread is IO Thread, setup may emit this error.

If that’s the case, I wonder in general if Spring has some way / annotation to set a block-able worker Thread to run the publishCloudEventSync() or configure that Netty callback runs on block-able worker Thread. @saragluna

**Comment by saragluna** [comment]
@1-javier-fernandez I am wondering how is the `publishCloudEventSync()` method called in your project? Do you have the codesnippet for that?

**Comment by jfggdl** [comment]
@saragluna It is exactly the code pasted on the top when I created the issue. Thanks for looking into this.

**Comment by saragluna** [comment]
Hi @jfggdl, but the `publishCloudEventSync` needs to be called, maybe by a controller?

**Comment by jfggdl** [comment]
Correct. It is a simple call to the publishCloudEventSync() from a controller method.

**Comment by jfggdl** [comment]
@saragluna do we have an ETA as to when this issue should be mitigated? Thanks. This is Javier from Event Grid, by the way. You may want to ping me in Teams to get my attention immediately.
